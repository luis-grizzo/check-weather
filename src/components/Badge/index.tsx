interface BadgeProps {
  children: string
}

export function Badge({ children }: BadgeProps) {
  return (
    <span className="flex items-center gap-2 px-2 py-1 text-xs bg-neutral-100/60 border-1 border-neutral-600/10 text-neutral-950 rounded-lg backdrop-blur">
      {children}
    </span>
  )
}
