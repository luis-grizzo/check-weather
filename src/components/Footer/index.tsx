'use client'

export function Footer() {
  const currentYear = new Date().getFullYear()

  return (
    <footer className="bg-neutral-100/60 border-t-1 border-neutral-600/10">
      <p className="container mx-auto px-8 py-6 text-xs text-center">
        Developed by{' '}
        <a
          href="https://luisgrizzo.dev"
          target="_blank"
          rel="noreferrer"
          className="text-blue-600 hover:text-blue-400 transition-colors"
        >
          Lu√≠s Grizzo
        </a>
        {', '}
        <a
          href="https://github.com/luis-grizzo/check-weather?tab=MIT-1-ov-file"
          target="_blank"
          className="text-xs text-blue-600 hover:text-blue-400 transition-colors"
          rel="noreferrer"
        >
          MIT License
        </a>
        {', '}
        {currentYear}.
      </p>
    </footer>
  )
}
