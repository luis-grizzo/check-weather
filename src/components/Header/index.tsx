'use client'

import { useEffect, useState } from 'react'

import { hourInMilliseconds } from '@/constants/timeMarks'

import {
  formatCountryName,
  formatDate,
  formatTime,
  formatTimer
} from '@/lib/stringFormaters'
import { positiveOrZero } from '@/lib/numberFormaters'

import type { FetchWeatherFactoredResponse } from '@/utils/weatherUtils'
import { useRouter } from 'next/navigation'

interface HeaderProps {
  location: FetchWeatherFactoredResponse['location']
  requestTimestamp: number
}

export function Header({ location, requestTimestamp }: HeaderProps) {
  const router = useRouter()

  const [countdownTimer, setCountdownTimer] = useState('--:--')
  const [shouldRefresh, setShouldRefresh] = useState(false)

  useEffect(() => {
    const timer = setInterval(() => {
      const currentTime = Date.now()
      const revalidateTime = requestTimestamp * 1_000 + hourInMilliseconds

      const remainTime = revalidateTime - currentTime
      const validatedRemainTime = positiveOrZero(remainTime)

      const formattedTimer = formatTimer(validatedRemainTime)

      setCountdownTimer(formattedTimer)

      if (!validatedRemainTime) {
        setShouldRefresh(true)

        clearInterval(timer)
      }
    }, 1_000)

    return () => clearInterval(timer)
  }, [requestTimestamp])

  useEffect(() => {
    if (shouldRefresh) {
      router.refresh()

      setShouldRefresh(false)
    }
  }, [shouldRefresh])

  return (
    <header className="flex items-center justify-between container mx-auto px-8 py-6">
      <div className="flex flex-col gap-1">
        <span role="timer" className="text-xs">
          {`Next forecast in ${countdownTimer}`}
        </span>

        {location.country && (
          <span className="text-lg">
            {`${location.city}, ${formatCountryName(location.country)}`}
          </span>
        )}

        <span className="text-xs">{`${formatTime(requestTimestamp)} - ${formatDate(requestTimestamp)}`}</span>
      </div>
    </header>
  )
}
