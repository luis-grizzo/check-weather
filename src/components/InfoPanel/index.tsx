import Image from 'next/image'

import { layoutIcons } from '@/constants/icons'

interface InfoPanelProps {
  children: React.ReactNode
}

export function InfoPanel({ children }: InfoPanelProps) {
  return (
    <div
      role="dialog"
      className="flex items-start gap-4 p-4 text-sm text-red-950 bg-red-100/60 border-red-600/10 border-1 rounded-xl"
    >
      <Image src={layoutIcons.error} alt="" className="w-5 aspect-square" />

      {children}
    </div>
  )
}
