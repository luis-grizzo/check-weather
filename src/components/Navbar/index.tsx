'use client'

import { useEffect, useState } from 'react'
import Image from 'next/image'
import { motion, AnimatePresence } from 'framer-motion'

import { weatherIconsArray } from '@/constants/icons'
import { calculateRandomIndex } from '@/lib/calculators'

export function Navbar() {
  const [stateIndex, setStateIndex] = useState(0)

  const MotionImage = motion(Image)

  useEffect(() => {
    const timer = setTimeout(() => {
      let newIndex = stateIndex

      while (newIndex === stateIndex)
        newIndex = calculateRandomIndex(weatherIconsArray)

      setStateIndex(newIndex)
    }, 5_000)

    return () => clearTimeout(timer)
  }, [stateIndex])

  return (
    <nav className="sticky top-0 min-h-[76px] bg-neutral-100/80 border-b-1 border-neutral-600/10 backdrop-blur overflow-hidden z-50">
      <div className="flex items-center gap-2 container mx-auto px-8 py-6">
        <AnimatePresence mode="wait">
          {weatherIconsArray.map(
            (icon, index) =>
              stateIndex === index && (
                <MotionImage
                  key={icon.alt}
                  role="img"
                  initial={{ opacity: 0, y: -32 }}
                  animate={{ opacity: 1, y: 0 }}
                  exit={{ opacity: 0, y: 32 }}
                  src={icon.src}
                  alt={icon.alt}
                  className="w-7 aspect-square"
                />
              )
          )}
        </AnimatePresence>

        <span className="text-lg">Check Weather</span>
      </div>
    </nav>
  )
}
