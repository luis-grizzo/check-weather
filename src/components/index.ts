export * from './Badge'
export * from './Button'
export * from './DisplayTime'
export * from './Footer'
export * from './Header'
export * from './InfoPanel'
export * from './Navbar'
export * from './RefreshButton'
