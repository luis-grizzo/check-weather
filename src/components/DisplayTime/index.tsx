'use client'

import { Badge } from '@/components/Badge'

import { formatTime } from '@/lib/stringFormaters'

interface DisplayTimeProps {
  description: string
  timestamp: number
}

export function DisplayTime({ description, timestamp }: DisplayTimeProps) {
  return <Badge>{`${description} ${formatTime(timestamp)}`}</Badge>
}
