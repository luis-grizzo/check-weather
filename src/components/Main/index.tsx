'use client'

import Image from 'next/image'

import { Badge } from '@/components/Badge'

import useWeather from '@/hooks/useWeather'
import { useLocation } from '@/hooks/useLocation'

import { formatDecimals } from '@/utils/formatDecimals'
import { roundValue } from '@/utils/roundValue'
import { formatTime } from '@/utils/formatTime'

import { weathers } from '@/constants/weathers'

import air from '@public/air.svg'
import water_drop from '@public/water_drop.svg'
import visibility from '@public/visibilit.svg'

export function Main() {
  const { coords } = useLocation()

  const { weather } = useWeather({
    latitude: formatDecimals(coords?.latitude),
    longitude: formatDecimals(coords?.longitude)
  })

  if (weather) {
    const isDay =
      weather.dt > weather.sys.sunrise && weather.dt < weather.sys.sunset

    const color =
      weathers[weather.weather[0].main.toLowerCase() as keyof typeof weathers]
        .color
    const icon =
      weathers[weather.weather[0].main.toLowerCase() as keyof typeof weathers]
        .icon[isDay ? 'day' : 'night']

    return (
      <main
        className={`relative flex flex-col items-center justify-center gap-8 h-full p-8 overflow-hidden before:absolute before:w-3/4 before:max-w-96 before:aspect-square before:bg-[radial-gradient(circle_at_center,_var(--tw-gradient-stops))] ${color} before:to-transparent before:to-70%  before:rounded-full before:animate-pulse before:-z-50`}
      >
        <div className="flex items-center gap-4">
          <Image src={icon} alt=" " className="w-24 aspect-square" />

          <span className="text-8xl">{roundValue(weather.main.temp)}째</span>
        </div>

        <div className="flex flex-wrap justify-center gap-2">
          <Badge>Fells Like {roundValue(weather.main.feels_like)}째</Badge>
          <Badge>Min {roundValue(weather.main.temp_min)}째</Badge>
          <Badge>Max {roundValue(weather.main.temp_max)}째</Badge>
          <Badge>Sunrise {formatTime(weather.sys.sunrise)}</Badge>
          <Badge>Sunset {formatTime(weather.sys.sunset)}</Badge>
        </div>

        <span className="text-base capitalize">
          {weather.weather[0].description}
        </span>

        <div className="flex items-center justify-between gap-8 p-8 bg-neutral-100/60 border-1 border-neutral-600/10 rounded-2xl backdrop-blur">
          <div className="flex flex-col items-center gap-2">
            <Image src={air} alt="" className="w-8 h-8" />

            <span className="text-lg text-nowrap">
              {roundValue(weather.wind.speed)} km/h
            </span>

            <span className="text-xs">Wind</span>
          </div>

          <div className="flex flex-col items-center gap-2">
            <Image src={water_drop} alt="" className="w-8 h-8" />

            <span className="text-lg text-nowrap">
              {weather.main.humidity}%
            </span>

            <span className="text-xs">Humidity</span>
          </div>

          <div className="flex flex-col items-center gap-2">
            <Image src={visibility} alt="" className="w-8 h-8" />

            <span className="text-lg text-nowrap">
              {weather.visibility / 1000} km
            </span>

            <span className="text-xs">Visibility</span>
          </div>
        </div>
      </main>
    )
  }
}
