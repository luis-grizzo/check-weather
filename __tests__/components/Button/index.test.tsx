import { Button } from '@/components'
import { render } from '@testing-library/react'

describe('Button', () => {
  it('should match snapshot', () => {
    const { container } = render(<Button>test</Button>)

    expect(container).toMatchSnapshot()
  })
})
