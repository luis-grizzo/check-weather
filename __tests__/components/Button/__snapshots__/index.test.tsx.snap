// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<div>
  <button
    class="flex items-center justify-center py-3 px-4 bg-neutral-100/60 border-1 border-neutral-600/10 rounded-lg hover:bg-neutral-200/60 transition-colors"
  >
    test
  </button>
</div>
`;
